lib/src/local_working_copy.rs

```diff
@@ -1177,24 +1176,23 @@ impl TreeState {
         Ok(FileState::for_file(executable, size, &metadata))
     }
 
-    #[cfg_attr(windows, allow(unused_variables))]
     fn write_symlink(&self, disk_path: &Path, target: String) -> Result<FileState, CheckoutError> {
-        #[cfg(windows)]
-        {
-            println!("ignoring symlink at {}", disk_path.display());
-        }
-        #[cfg(unix)]
-        {
-            let target = PathBuf::from(&target);
-            symlink(&target, disk_path).map_err(|err| CheckoutError::Other {
-                message: format!(
-                    "Failed to create symlink from {} to {}",
-                    disk_path.display(),
-                    target.display()
-                ),
-                err: err.into(),
-            })?;
-        }
+        let target = PathBuf::from(&target);
+        try_symlink(&target, disk_path, Path::is_dir(disk_path)).map_err(|inner_err| {
```

Comment:
maybe you mean something like `disk_path.parent().join(target).is_dir()`?

(But see also the comment in the previous discussion https://github.com/martinvonz/jj/pull/2939#discussion_r1481542573)

Reply:
Isn't it the function to create the `disk_path` link? Also, the `target` directory/file might not be materialized yet. See https://github.com/martinvonz/jj/pull/2939#discussion_r1481542573.

Reply:
I think I see the cause of the confusion. I’ve misunderstood Path::is_dir(); I didn’t think it did I/O. My goal was to *calculate* the directory-ness of the path at which the symlink should be created, since (as you say) neither the link nor its target may exist on disk yet; I see now that isn’t possible. 

Here’s a backup plan which I’ll implement in another revision of the branch: check whether the symlink’s target is a repo-relative path; if so, and it’s in the tree, it must be a file. If it’s relative but not in the tree, it’s a directory. If it points outside the tree (can this happen?), we can stat the target. If it points outside the tree but to a file or directory that doesn’t exist… I don’t know. An error saying “please don’t create symlinks to paths outside the repository”??

Reply:
> check whether the symlink’s target is a repo-relative path; if so, and it’s in the tree, it must be a file. If it’s relative but not in the tree, it’s a directory. 

nit: if it's a symlink, check recursively (with some cycle detection.)

> If it points outside the tree, we could stat the target.

I would avoid stating external files because it's scary if `jj` read arbitrary unmanaged data (though leaking file node type wouldn't be a security issue.) I think it's okay to fall back to a file symlink (or a normal text file) in that case.

BTW, in the initial implementation, I think it's okay to always assume symlinks as file links. The user can fix it up manually so long as `jj` can read both file and directory symlinks.